src/main/scala/vsys/blockchain/state/opcdiffs/BasicOpcDiff.scala

```diff
@@ -4,109 +4,137 @@ import vsys.blockchain.transaction.ValidationError
 
   object BasicType extends Enumeration {
-    sealed case class basicTypeVal(
+    sealed case class BasicTypeVal(
       basicType: Int,
       len: Int,
-      differ: (Array[Byte], Seq[DataEntry]) => Either[ValidationError, Seq[DataEntry]])
+      indexes: Seq[Int],
+      op: (Array[Byte], Seq[DataEntry]) => Either[ValidationError, DataEntry])
     extends Val(basicType) { def *(n: Int): Int = n * basicType }
 
-    val Add         = basicTypeVal(1, 4, (b, d) => operation(d(b(1)), d(b(2)), d, b(3), add))
-    val Minus       = basicTypeVal(2, 4, (b, d) => operation(d(b(1)), d(b(2)), d, b(3), minus))
-    val Multiply    = basicTypeVal(3, 4, (b, d) => operation(d(b(1)), d(b(2)), d, b(3), multiply))
-    val Divide      = basicTypeVal(4, 4, (b, d) => operation(d(b(1)), d(b(2)), d, b(3), divide))
-    val Minimum     = basicTypeVal(5, 4, (b, d) => operation(d(b(1)), d(b(2)), d, b(3), minimum))
-    val Maximum     = basicTypeVal(6, 4, (b, d) => operation(d(b(1)), d(b(2)), d, b(3), maximum))
-    val Concat      = basicTypeVal(7, 4, (b, d) => concat(d(b(1)), d(b(2)), d, b(3)))
-    val ConstantGet = basicTypeVal(8, 2, (b, d) => constantGet(b.slice(1, b.length-1), d, b(b.length-1)))
-    val SqrtBigInt  = basicTypeVal(9, 3, (b, d) => sqrt(d(b(1)), d, b(2)))
+    val Add         = BasicTypeVal(1,  4, Seq(1, 2), (b, d) => numBiOperation(d(b(1)), d(b(2)), add))
+    val Minus       = BasicTypeVal(2,  4, Seq(1, 2), (b, d) => numBiOperation(d(b(1)), d(b(2)), minus))
+    val Multiply    = BasicTypeVal(3,  4, Seq(1, 2), (b, d) => numBiOperation(d(b(1)), d(b(2)), multiply))
+    val Divide      = BasicTypeVal(4,  4, Seq(1, 2), (b, d) => numBiOperation(d(b(1)), d(b(2)), divide))
+    val Minimum     = BasicTypeVal(5,  4, Seq(1, 2), (b, d) => numBiOperation(d(b(1)), d(b(2)), minimum))
+    val Maximum     = BasicTypeVal(6,  4, Seq(1, 2), (b, d) => numBiOperation(d(b(1)), d(b(2)), maximum))
+    val Concat      = BasicTypeVal(7,  4, Seq(1, 2), (b, d) => concat(d(b(1)), d(b(2))))
+    val ConstantGet = BasicTypeVal(8,  2, Seq(),     (b, d) => constantGet(b.slice(1, b.length-1)))
+    val SqrtBigInt  = BasicTypeVal(9,  3, Seq(1),    (b, d) => sqrt(d(b(1))))
+    val Convert     = BasicTypeVal(10, 4, Seq(1, 2), (b, d) => convertion(d(b(1)), d(b(2))))
+    val And         = BasicTypeVal(11, 4, Seq(1, 2), (b, d) => boolBiOperation(d(b(1)), d(b(2)), bAnd))
+    val Or          = BasicTypeVal(12, 4, Seq(1, 2), (b, d) => boolBiOperation(d(b(1)), d(b(2)), bOr))
+    val Xor         = BasicTypeVal(13, 4, Seq(1, 2), (b, d) => boolBiOperation(d(b(1)), d(b(2)), bXor))
+    val Not         = BasicTypeVal(14, 3, Seq(1),    (b, d) => not(d(b(1))))
   }
 
+  private def formatRes[T] (res: T, dt: DataTypeVal[T]): Either[ValidationError, DataEntry] =
+    formatResB[T](dt.serializer(res), dt)
+
+  private def formatResB[T] (res: Array[Byte], dt: DataTypeVal[T]): Either[ValidationError, DataEntry] =
+    DataEntry.create(res, dt).left.map(_ => ValidationError.OverflowError)
```

Comment:
When will it return `ValidationError.OverflowError`?

Reply:
For example, BigInt * BigInt, an InvalidDataEntry will be returned by `DataEntry.create`, but actually it is an Overflow.

Reply:
I see. So the usage here should be careful that we will only return `ValidationError.OverflowError` and `DataEntry`